### PR TITLE
CNS-362: Unitests subscriptions

### PR DESCRIPTION
**Pull request by oren-lava**
Added some unit tests for the subscription functionality and its integration with the pairing and projects modules.

### CODE REVIEW

File: x/pairing/keeper/pairing_subscription_test.go [hunk 1]
```diff
@@ -121,42 +121,479 @@ func TestRelayPaymentSubscription(t *testing.T) {
...
+	}
+}
+
+func TestStrictestPolicyProvidersToPair(t *testing.T) {
+	ts := setupForPaymentTest(t)
+
+	// add 5 more providers so we can have enough providers for testing
+	ts.addProvider(5)
+
+	err := ts.keepers.Subscription.CreateSubscription(sdk.UnwrapSDKContext(ts.ctx),
```
Comment (Yaroms) [inline x/pairing/keeper/pairing_subscription_test.go:281]: use msg server

File: x/pairing/keeper/pairing_subscription_test.go [hunk 2]
```diff
@@ -121,42 +121,479 @@ func TestRelayPaymentSubscription(t *testing.T) {
...
+			adminPolicy := &projectstypes.Policy{
+				GeolocationProfile: 1,
+				MaxProvidersToPair: tt.providersToPairAdminPolicy,
+			}
+			subscriptionPolicy := &projectstypes.Policy{
+				GeolocationProfile: 1,
+				MaxProvidersToPair: tt.providersToPairSubPolicy,
+			}
+
+			err = ts.keepers.Projects.SetPolicy(sdk.UnwrapSDKContext(ts.ctx), []string{proj.Index}, adminPolicy,
```
Comment (Yaroms) [inline x/pairing/keeper/pairing_subscription_test.go:313]: use msg server

**Review by Yaroms — CHANGES_REQUESTED** [review]
try to use the msg server wherever you can instead of using the keepers directly

File: x/subscription/keeper/subscription_test.go [hunk 3]
```diff
@@ -292,15 +292,35 @@ func TestRenewSubscription(t *testing.T) {
...
 	sub, found = keeper.GetSubscription(ts.ctx, creator)
 	require.True(t, found)
 
-	require.Equal(t, uint64(13), sub.DurationLeft)
-	require.Equal(t, uint64(10), sub.DurationTotal)
+	require.Equal(t, uint64(12), sub.DurationLeft)
+	require.Equal(t, uint64(9), sub.DurationTotal)
+
+	// edit the subscription's plan (allow more CU)
+	subPlan, found := ts.keepers.Plans.FindPlan(ts.ctx, sub.PlanIndex, sub.PlanBlock)
```
[the remark this answers is not in the document]
Comment (oren-lava) [reply x/subscription/keeper/subscription_test.go:306]: done

File: x/subscription/keeper/subscription_test.go [hunk 4]
```diff
@@ -292,15 +292,35 @@ func TestRenewSubscription(t *testing.T) {
...
+
+	// try extending the subscription (normally we could extend with 1 more month, but since the
+	// subscription's plan changed, the extension should fail)
+	err = keeper.CreateSubscription(ts.ctx, creator, creator, ts.plans[0].Index, 1, "")
+	require.NotNil(t, err)
+	require.Equal(t, uint64(12), sub.DurationLeft)
+	require.Equal(t, uint64(9), sub.DurationTotal)
+
+	// get the subscription's plan and make sure it uses the old plan
+	subPlan, found = ts.keepers.Plans.FindPlan(ts.ctx, sub.PlanIndex, sub.PlanBlock)
```
[the remark this answers is not in the document]
Comment (oren-lava) [reply x/subscription/keeper/subscription_test.go:321]: As we talked, in this particular place I cannot use the query server since the query always gets the latest plan version (and here I need an older one to verify that the sub still uses it)

File: x/pairing/keeper/pairing_subscription_test.go [hunk 5]
```diff
@@ -121,42 +121,479 @@ func TestRelayPaymentSubscription(t *testing.T) {
...
+	for _, tt := range geolocationTestTemplates {
+		t.Run(tt.name, func(t *testing.T) {
+			adminPolicy := &projectstypes.Policy{
+				GeolocationProfile: tt.geolocationAdminPolicy,
+			}
+			subscriptionPolicy := &projectstypes.Policy{
+				GeolocationProfile: tt.geolocationSubPolicy,
+			}
+
+			err = ts.keepers.Projects.SetPolicy(sdk.UnwrapSDKContext(ts.ctx), []string{proj.Index}, adminPolicy,
```
[the remark this answers is not in the document]
Comment (oren-lava) [reply x/pairing/keeper/pairing_subscription_test.go:248]: done

File: x/pairing/keeper/pairing_subscription_test.go (hunk 1 above)
Comment (oren-lava) [reply x/pairing/keeper/pairing_subscription_test.go:281]: done

File: x/pairing/keeper/pairing_subscription_test.go (hunk 2 above)
Comment (oren-lava) [reply x/pairing/keeper/pairing_subscription_test.go:313]: done

File: x/projects/keeper/project_test.go [hunk 6]
```diff
@@ -357,3 +436,126 @@ func SetPolicyTest(t *testing.T, testAdminPolicy bool) {
...
+	subAccount2 := common.CreateNewAccount(ctx, *keepers, 10000)
+	devAcc := common.CreateNewAccount(ctx, *keepers, 10000)
+
+	projectName1 := "mockname1"
+	projectName2 := "mockname2"
+
+	projectID1 := types.ProjectIndex(subAccount1.Addr.String(), projectName1)
+	projectID2 := types.ProjectIndex(subAccount2.Addr.String(), projectName2)
+
+	projectData1 := types.ProjectData{
```
Comment (orenl-lava) [inline x/projects/keeper/project_test.go:505]: can you reuse or enhance the existing `prepareProjectsData()`?

File: x/subscription/keeper/subscription_test.go [hunk 7]
```diff
@@ -292,15 +292,35 @@ func TestRenewSubscription(t *testing.T) {
 	err = keeper.CreateSubscription(ts.ctx, creator, creator, ts.plans[0].Index, 12, "")
 	require.NotNil(t, err)
 
-	// but asking for additional 10 is fine
-	err = keeper.CreateSubscription(ts.ctx, creator, creator, ts.plans[0].Index, 10, "")
+	// but asking for additional 9 months (10 would also be fine)
```
Comment (orenl-lava) [inline x/subscription/keeper/subscription_test.go:295]: why this change?
the intent here was to ask for the max possible.

File: x/subscription/keeper/subscription_test.go [hunk 8]
```diff
@@ -327,63 +347,107 @@ func TestMonthlyRechargeCU(t *testing.T) {
...
+	anotherAccount := common.CreateNewAccount(ts._ctx, *ts.keepers, 10000)
 	creator := account.Addr.String()
 
-	err := keeper.CreateSubscription(ts.ctx, creator, creator, ts.plans[0].Index, 2, "")
+	err := keeper.CreateSubscription(ts.ctx, creator, creator, ts.plans[0].Index, 3, "")
 	require.Nil(t, err)
 
-	block1 := uint64(ts.ctx.BlockHeight())
+	// add another project under the subcscription
+	projectData := projectstypes.ProjectData{
```
Comment (orenl-lava) [inline x/subscription/keeper/subscription_test.go:357]: can you reuse or enhance `prepareProjectsData()` here?

File: x/subscription/keeper/subscription_test.go [hunk 9]
```diff
@@ -518,9 +582,12 @@ func TestAddProjectToSubscription(t *testing.T) {
 
 	defaultProjectName := projectstypes.ADMIN_PROJECT_NAME
 	longProjectName := strings.Repeat(defaultProjectName, projectstypes.MAX_PROJECT_NAME_LEN)
+	projectNameWithComma := "projectName,"
```
Comment (orenl-lava) [inline x/subscription/keeper/subscription_test.go:585]: enough to have one test with disallowed character?
(tests for checking different disallowed characters should go with the helper that checks names/descriptions)

**Review by orenl-lava — CHANGES_REQUESTED** [review]

File: x/pairing/keeper/pairing.go [hunk 10]
```diff
@@ -198,7 +198,10 @@ func (k Keeper) CalculateEffectiveProvidersToPairFromPolicies(policies []*projec
 
 	for _, policy := range policies {
```
[the remark this answers is not in the document]
Comment (oren-lava) [reply x/pairing/keeper/pairing.go:199]: yeah but I wanted to use the findmin function. Does it make a big difference?

[the remark this answers is not in the document]
Comment (oren-lava) [reply x/pairing/keeper/pairing.go:199]: the findmin function returns zero for an empty array. So if all the policies are nil, we'll get providersToPair = 0. I'll add some code that will trigger an error when providersToPair = 0.

File: x/pairing/keeper/pairing.go [hunk 11]
```diff
@@ -198,7 +198,10 @@ func (k Keeper) CalculateEffectiveProvidersToPairFromPolicies(policies []*projec
...
 		if policy != nil {
```
[the remark this answers is not in the document]
Comment (oren-lava) [reply x/pairing/keeper/pairing.go:200]: the findmin function returns zero for an empty array. So if all the policies are nil, we'll get providersToPair = 0. I'll add some code that will trigger an error when providersToPair = 0. Also, the plan policy will never be nil (still adding the error trigger code though)

File: x/projects/keeper/creation.go [hunk 12]
```diff
@@ -82,6 +82,12 @@ func (k Keeper) RegisterKey(ctx sdk.Context, key types.ProjectKey, project *type
 					return utils.LavaError(ctx, k.Logger(ctx), "RegisterKey_add_dev_key_failed", details, "adding developer key failed")
 				}
 			}
+		default:
```
[the remark this answers is not in the document]
Comment (oren-lava) [reply x/projects/keeper/creation.go:85]: yeah, it shouldn't happen since it's checked when the TX is received (just noticed I didn't check this in the addProject TX, so I've added a check). So what do you suggest? omit this case? panic? keep as is?

File: x/projects/keeper/project_test.go [hunk 13]
```diff
@@ -57,11 +57,66 @@ func TestCreateProject(t *testing.T) {
 
 	testkeeper.AdvanceEpoch(ctx, keepers)
 
-	// create another project with the same name, should fail as this is unique
-	err = keepers.Projects.CreateProject(sdk.UnwrapSDKContext(ctx), subAccount.Addr.String(), projectData, plan)
+	// test invalid project name/description
+	defaultProjectName := types.ADMIN_PROJECT_NAME
+	longProjectName := strings.Repeat(defaultProjectName, types.MAX_PROJECT_NAME_LEN)
```
[the remark this answers is not in the document]
Comment (oren-lava) [reply x/projects/keeper/project_test.go:62]: it doesn't matter since the defaultProjectName's length is bigger than 1. Assuming MAX_PROJECT_NAME_LEN = 50 and defaultProjectName = "admin" then len(longProjectName) = 50*5 which is way more than 50. But I'll add the +1 for readability

File: x/projects/keeper/project_test.go [hunk 14]
```diff
@@ -57,11 +57,66 @@ func TestCreateProject(t *testing.T) {
...
+	projectNameWithComma := "projectName,"
+	nonAsciiProjectName := "projectName¢"
+
+	projectDescription := "test project"
+	longProjectDescription := strings.Repeat(projectDescription, types.MAX_PROJECT_DESCRIPTION_LEN)
```
[the remark this answers is not in the document]
Comment (oren-lava) [reply x/projects/keeper/project_test.go:67]: done

File: x/projects/keeper/project_test.go [hunk 15]
```diff
@@ -57,11 +57,66 @@ func TestCreateProject(t *testing.T) {
 
 	testkeeper.AdvanceEpoch(ctx, keepers)
 
-	// create another project with the same name, should fail as this is unique
-	err = keepers.Projects.CreateProject(sdk.UnwrapSDKContext(ctx), subAccount.Addr.String(), projectData, plan)
+	// test invalid project name/description
+	defaultProjectName := types.ADMIN_PROJECT_NAME
+	longProjectName := strings.Repeat(defaultProjectName, types.MAX_PROJECT_NAME_LEN)
+	projectNameWithComma := "projectName,"
```
[the remark this answers is not in the document]
Comment (oren-lava) [reply x/projects/keeper/project_test.go:63]: Do you mean only checking projectNameWithComma? why? I want to make sure that all the restrictions in the validateProjectName are tested (so if someone in the future will change something, we'll know)

File: x/projects/keeper/project_test.go [hunk 16]
```diff
@@ -57,11 +57,66 @@ func TestCreateProject(t *testing.T) {
...
+		{"bad projectName (contains comma)", projectNameWithComma, projectDescription},
+		{"bad projectName (non ascii)", nonAsciiProjectName, projectDescription},
+		{"bad projectName (empty)", "", projectDescription},
+		{"bad projectDescription (too long)", "test1", longProjectDescription},
+		{"bad projectDescription (non ascii)", "test2", nonAsciiProjectDescription},
+	}
+
+	for _, tt := range nameAndDescriptionTests {
+		t.Run(tt.name, func(t *testing.T) {
+			nameDescriptionTestProjectData.Name = tt.projectName
```
[the remark this answers is not in the document]
Comment (oren-lava) [reply x/projects/keeper/project_test.go:89]: projectData exists. changed it to testProjectData

File: x/projects/keeper/project_test.go [hunk 17]
```diff
@@ -57,11 +57,66 @@ func TestCreateProject(t *testing.T) {
...
+
+	// continue testing traits that are not related to the project's name/description
+	// try creating a project with invalid project keys
+	invalidKeysProjectData := projectData
+	invalidKeysProjectData.Name = "nonDuplicateProjectName"
+	invalidKeysProjectData.ProjectKeys = []types.ProjectKey{
+		{
+			Key:   subAccount.Addr.String(),
+			Types: []types.ProjectKey_KEY_TYPE{types.ProjectKey_DEVELOPER},
+			Vrfpk: "",
```
[the remark this answers is not in the document]
Comment (oren-lava) [reply x/projects/keeper/project_test.go:105]: The subscription account is automatically a developer when the admin project is created. When a normal project is created (using the createProject function) it's not true, so this code is necessary

File: x/projects/keeper/project_test.go [hunk 18]
```diff
@@ -57,11 +57,66 @@ func TestCreateProject(t *testing.T) {
...
+		},
+		{
+			Key:   adminAcc.Addr.String(),
+			Types: []types.ProjectKey_KEY_TYPE{4},
```
[the remark this answers is not in the document]
Comment (oren-lava) [reply x/projects/keeper/project_test.go:109]: it shouldn't happen because we have checks for that. But, I don't think there's harm in keeping it. Am I wrong?

File: x/projects/keeper/project_test.go (hunk 17 above)
[the remark this answers is not in the document]
Comment (oren-lava) [reply x/projects/keeper/project_test.go:105]: are you refering to the vrfpk field? if so, it'll be fixed when I'll handle the merge conflicts

File: x/projects/keeper/project_test.go [hunk 19]
```diff
@@ -57,11 +57,66 @@ func TestCreateProject(t *testing.T) {
...
+			Vrfpk: "",
```
[the remark this answers is not in the document]
Comment (oren-lava) [reply x/projects/keeper/project_test.go:110]: see comment above

File: x/projects/keeper/project_test.go [hunk 20]
```diff
@@ -242,41 +304,58 @@ func SetPolicyTest(t *testing.T, testAdminPolicy bool) {
...
+
+		{
+			"bad projectID (doesn't exist)", developerAcc.Addr.String(), "fakeProjectId", uint64(1),
 			[]types.ChainPolicy{{ChainId: spec.Index, Apis: []string{spec.Apis[0].Name}}},
 			100, 10, 3, true, false,
 		},
 
 		{
-			"bad chainID (doesn't exist)", subAccount.Addr.String(),
+			"invalid geolocation (0)", developerAcc.Addr.String(), "fakeProjectId", uint64(0),
```
[the remark this answers is not in the document]
Comment (oren-lava) [reply x/projects/keeper/project_test.go:331]: not sure I understand your comment

File: x/projects/keeper/project_test.go [hunk 21]
```diff
@@ -221,19 +280,22 @@ func SetPolicyTest(t *testing.T, testAdminPolicy bool) {
 	err := keepers.Projects.CreateProject(sdk.UnwrapSDKContext(ctx), subAccount.Addr.String(), projectData, plan)
 	require.Nil(t, err)
 
-	keepers.Projects.AddKeysToProject(sdk.UnwrapSDKContext(ctx), projectID, adminAcc.Addr.String(),
+	err = keepers.Projects.AddKeysToProject(sdk.UnwrapSDKContext(ctx), projectID, adminAcc.Addr.String(),
```
[the remark this answers is not in the document]
Comment (oren-lava) [reply x/projects/keeper/project_test.go:283]: 1. done
2. I did see that in your code you use _ctx. But this is problematic when you need to update the context in your test (like when you advance an epoch) since _ctx is created in the beginning of the code, before any context changes. Since this test advances blocks, I won't use _ctx

File: x/projects/keeper/project_test.go [hunk 22]
```diff
@@ -357,3 +436,126 @@ func SetPolicyTest(t *testing.T, testAdminPolicy bool) {
...
+	plan := common.CreateMockPlan()
+
+	projectData := types.ProjectData{
+		Name:        projectName,
+		Description: "",
+		Enabled:     true,
+		ProjectKeys: []types.ProjectKey{{
+			Key:   subAccount.Addr.String(),
+			Types: []types.ProjectKey_KEY_TYPE{types.ProjectKey_DEVELOPER},
+			Vrfpk: "",
```
[the remark this answers is not in the document]
Comment (oren-lava) [reply x/projects/keeper/project_test.go:457]: will be handled in the merge conflict resolution

File: x/projects/keeper/project_test.go [hunk 23]
```diff
@@ -357,3 +436,126 @@ func SetPolicyTest(t *testing.T, testAdminPolicy bool) {
 		})
 	}
 }
+
+func TestAddDevKeyToSameProjectDifferentBlocks(t *testing.T) {
+	_, keepers, ctx := testkeeper.InitAllKeepers(t)
+
+	projectName := "mockname1"
+	subAccount := common.CreateNewAccount(ctx, *keepers, 10000)
```
[the remark this answers is not in the document]
Comment (oren-lava) [reply x/projects/keeper/project_test.go:444]: done

File: x/projects/keeper/project_test.go [hunk 24]
```diff
@@ -357,3 +436,126 @@ func SetPolicyTest(t *testing.T, testAdminPolicy bool) {
...
+	err := keepers.Projects.CreateProject(sdk.UnwrapSDKContext(ctx), subAccount.Addr.String(), projectData, plan)
+	require.Nil(t, err)
+
+	ctx = testkeeper.AdvanceBlock(ctx, keepers)
+
+	err = keepers.Projects.AddKeysToProject(sdk.UnwrapSDKContext(ctx), projectID, subAccount.Addr.String(),
+		[]types.ProjectKey{{
+			Key:   devAcc1.Addr.String(),
+			Types: []types.ProjectKey_KEY_TYPE{types.ProjectKey_DEVELOPER},
+			Vrfpk: "",
```
[the remark this answers is not in the document]
Comment (oren-lava) [reply x/projects/keeper/project_test.go:470]: will be handled in the merge conflict resolution

File: x/projects/keeper/project_test.go [hunk 25]
```diff
@@ -357,3 +436,126 @@ func SetPolicyTest(t *testing.T, testAdminPolicy bool) {
...
+		}})
+	require.Nil(t, err)
+
+	ctx = testkeeper.AdvanceBlock(ctx, keepers)
+
+	err = keepers.Projects.AddKeysToProject(sdk.UnwrapSDKContext(ctx), projectID, subAccount.Addr.String(),
+		[]types.ProjectKey{{
+			Key:   devAcc2.Addr.String(),
+			Types: []types.ProjectKey_KEY_TYPE{types.ProjectKey_DEVELOPER},
+			Vrfpk: "",
```
[the remark this answers is not in the document]
Comment (oren-lava) [reply x/projects/keeper/project_test.go:480]: will be handled in the merge conflict resolution

File: x/projects/keeper/project_test.go [hunk 26]
```diff
@@ -357,3 +436,126 @@ func SetPolicyTest(t *testing.T, testAdminPolicy bool) {
...
+	require.Nil(t, err)
+
+	require.Equal(t, 3, len(proj.ProjectKeys))
+}
+
+func TestAddDevKeyToDifferentProjectsInSameBlock(t *testing.T) {
+	_, keepers, ctx := testkeeper.InitAllKeepers(t)
+	plan := common.CreateMockPlan()
+
+	subAccount1 := common.CreateNewAccount(ctx, *keepers, 10000)
```
[the remark this answers is not in the document]
Comment (oren-lava) [reply x/projects/keeper/project_test.go:495]: done

File: x/projects/keeper/project_test.go (hunk 6 above)
Comment (oren-lava) [reply x/projects/keeper/project_test.go:505]: there is no such function?

File: x/subscription/keeper/subscription_test.go (hunk 7 above)
Comment (oren-lava) [reply x/subscription/keeper/subscription_test.go:295]: on line 315, I try to extend the subscription for another month. I wanted to create a situation where the subscription extension would succeed if everything was ok but I still expect a failure since the plan changed (if I would've kept the 10 month extension, another 1 month extension would fail whether the plan changed or not)

File: x/subscription/keeper/subscription_test.go (hunk 8 above)
Comment (oren-lava) [reply x/subscription/keeper/subscription_test.go:357]: there is no such function?

File: x/pairing/keeper/pairing.go [hunk 27]
```diff
@@ -198,7 +198,10 @@ func (k Keeper) CalculateEffectiveProvidersToPairFromPolicies(policies []*projec
...
-			providersToPairValues = append(providersToPairValues, policy.GetMaxProvidersToPair())
+			providersToPair := policy.GetMaxProvidersToPair()
+			if providersToPair > 1 {
```
[the remark this answers is not in the document]
Comment (oren-lava) [reply x/pairing/keeper/pairing.go:202]: removed the check here and moved the basic checks of add-project-keys and set-policy TXs to the msgServer (from the ValidateBasic function that is only called through the client code)

File: x/subscription/keeper/subscription_test.go (hunk 9 above)
Comment (oren-lava) [reply x/subscription/keeper/subscription_test.go:585]: see answer to same comment in project_test

File: x/projects/keeper/project_test.go (hunk 6 above)
Comment (orenl-lava) [reply x/projects/keeper/project_test.go:505]: it was merged on May 1, and may bring merge-conflict mayhem to you.
you should use/enhance it for everywhere you prepare a "types.ProjectData{ ... }"

File: x/subscription/keeper/subscription_test.go (hunk 7 above)
Comment (orenl-lava) [reply x/subscription/keeper/subscription_test.go:295]: got it. so maybe append to the comment: ", but one more is coming below".

**Review by orenl-lava — CHANGES_REQUESTED** [review]

File: x/pairing/keeper/pairing.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (orenl-lava) [reply x/pairing/keeper/pairing.go]: yes. our `FindMin()` is useful when you have a slice. it is awkward to use a loop to build a list for that (if you use a loop - then calculate ad-hoc - code is simpler, shorter, clearer). it is also inefficient (memory alloc/dealloc, and then a second loop in `FindMin()`).

File: x/projects/keeper/project_test.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (orenl-lava) [reply x/projects/keeper/project_test.go]: indeed both tests should trigger the error, but the one with LEN+1 is tighter and therefore better: if a bug in the validation allows for LEN+1 then the current test won't detect it.

File: x/projects/keeper/project_test.go (hunk 20 above)
[the remark this answers is not in the document]
Comment (orenl-lava) [reply x/projects/keeper/project_test.go]: oops... regex syntax to say: substitute the **"fakeProjectId"** with **projectID**.
otherwise the test will fail as planned because of bad projectID and not because of invalid geolocation as needed.

File: x/projects/keeper/project_test.go (hunk 21 above)
[the remark this answers is not in the document]
Comment (orenl-lava) [reply x/projects/keeper/project_test.go]: as you may see in the code, it re-assigns "ctx" when "_ctx" is updated, as needed.
in fact, I would love the testutils to have this convention all over, and have the functions return both "_ctx" and "ctx"!

File: x/projects/keeper/creation.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (oren-lava) [reply x/projects/keeper/creation.go:85]: done

File: x/projects/keeper/project_test.go (hunk 17 above)
[the remark this answers is not in the document]
Comment (oren-lava) [reply x/projects/keeper/project_test.go:106]: deleted it

File: x/subscription/keeper/subscription_test.go (hunk 9 above)
Comment (oren-lava) [reply x/subscription/keeper/subscription_test.go:585]: done

File: x/projects/keeper/project_test.go [hunk 28]
```diff
@@ -57,11 +57,66 @@ func TestCreateProject(t *testing.T) {
...
+			Vrfpk: "",
+		},
+		{
+			Key:   adminAcc.Addr.String(),
+			Types: []types.ProjectKey_KEY_TYPE{4},
+			Vrfpk: "",
+		},
+	}
+
+	// should fail because there's an invalid key
```
[the remark this answers is not in the document]
Comment (oren-lava) [reply x/projects/keeper/project_test.go:115]: done

**Comment by oren-lava** [comment]
closing this PR because I've transferred my work to another branch that originated from the latest main (to avoid lengthy conflict resolution)
